awx/main/tasks.py

```diff
@@ -3137,3 +3009,212 @@ def deep_copy_model_obj(
             permission_check_func(creater, copy_mapping.values())
     if isinstance(new_obj, Inventory):
         update_inventory_computed_fields.delay(new_obj.id)
+
+
+class AWXReceptorJob:
+    def __init__(self, task, runner_params):
+        self.task = task
+        self.runner_params = runner_params
+
+        if not self.task.instance.is_container_group_task:
+            execution_environment_params = self.task.build_execution_environment_params(self.task.instance)
+            self.runner_params['settings'].update(execution_environment_params)
+
+    def run(self):
+        # Create a socketpair. Where the left side will be used for writing our payload
+        # (private data dir, kwargs). The right side will be passed to Receptor for
+        # reading.
+        sockin, sockout = socket.socketpair()
+
+        threading.Thread(target=self.transmit, args=[sockin]).start()
```

Comment:
You never do a `.join()` on this do you?

Maybe I'm going to far with this, but some thoughts:

 - it doesn't seem to make sense to start `self.processor` if that thread is still running, so maybe a `.join()` on the transmit thread would be good to have assured before processing starts. Otherwise there's nothing to process.
 - We do need to run the cancel watcher while this is running, because it could get canceled while a large amount of data is being transmitted. Maybe we could poll the cancel callback until that transmit thread finishes?